conan/tools/files/patches.py

```diff
@@ -37,8 +38,10 @@ def patch(conanfile, base_path=None, patch_file=None, patch_string=None,
     :param kwargs: Extra parameters that can be added and will contribute to output information
     """
 
+    base_path = base_path or conanfile.source_folder
```

Comment:
This might have some risk.

Reply:
This makes patching depend on `no_copy_source`.
I think the behavior of `tools.patch` we want to see is to apply the patches relative to the working directory.

`apply_conandata_patches` can contain some logic to patch sources relative to the base source folder.

Reply:
`source_folder` always points to the right location, if `no_copy_source` if points to source, otherwise it points to `build_folder`

Reply:
So this will break:
```python
class CC(ConanFile):
    no_copy_source = True
    def build(self):
        with tools.chdir(self.build_folder):
            tools.patch("my-patch.patch")
```

Reply:
yes, because the patches are not in the build folder.

Reply:
What is the root of the ``base_path``? Shouldn't it be relative to some folder of the layout?
Recall this ``conan.tools.files patches`` is experimental, not used yet, we can change interface and behavior here, not a huge risk.